File: sandbox_template/build-template.ts
import { Template, defaultBuildLogger } from "e2b"

const TEMPLATE_NAME = "astraflow-desktop"
const BASE_TEMPLATE = "code-interpreter-v1"
const DEFAULT_DOMAIN = "cn-wlcb.sandbox.ucloudai.com"

function normalizeDomain(value: string | undefined) {
  const trimmed = value?.trim()

  if (!trimmed) {
    return undefined
  }

  return trimmed
    .replace(/^https?:\/\//i, "")
    .replace(/^\*\./, "")
    .replace(/\/+$/, "")
}

function normalizeUrl(value: string | undefined) {
  const trimmed = value?.trim()

  return trimmed || undefined
}

const apiKey = process.env.E2B_API_KEY
const domain = normalizeDomain(
  process.env.E2B_DOMAIN ??
    process.env.ASTRAFLOW_SANDBOX_DOMAIN ??
    DEFAULT_DOMAIN
)
const apiUrl = normalizeUrl(
  process.env.E2B_API_URL ?? process.env.ASTRAFLOW_SANDBOX_API_URL
)
const sandboxUrl = normalizeUrl(
  process.env.E2B_SANDBOX_URL ?? process.env.ASTRAFLOW_SANDBOX_URL
)
const validateApiKey = process.env.E2B_VALIDATE_API_KEY === "true"

const template = Template()
  .fromTemplate(BASE_TEMPLATE)
  .aptInstall(["tmux"], { noInstallRecommends: true })

const result = await Template.build(template, TEMPLATE_NAME, {
  apiKey,
  domain,
  apiUrl,
  sandboxUrl,
  validateApiKey,
  tags: ["latest"],
  onBuildLogs: defaultBuildLogger({ minLevel: "info" }),
})

console.log(
  JSON.stringify(
    {
      name: result.name,
      alias: result.alias,
      templateId: result.templateId,
      buildId: result.buildId,
      tags: result.tags,
    },
    null,
    2
  )
)

File: sandbox_template/python/build_template.py
import json
import os

from ucloud_sandbox import Template, default_build_logger


TEMPLATE_NAME = "astraflow-desktop"
BASE_TEMPLATE = "code-interpreter-v1"
DEFAULT_DOMAIN = "cn-wlcb.sandbox.ucloudai.com"
DEFAULT_CPU_COUNT = 2
DEFAULT_MEMORY_MB = 4096


def normalize_domain(value: str | None) -> str | None:
    if not value:
        return None

    trimmed = value.strip()
    if not trimmed:
        return None

    return (
        trimmed.removeprefix("https://")
        .removeprefix("http://")
        .removeprefix("*.")
        .rstrip("/")
    )


def read_positive_int(name: str, fallback: int) -> int:
    raw_value = os.environ.get(name, "").strip()
    if not raw_value:
        return fallback

    try:
        value = int(raw_value)
    except ValueError:
        return fallback

    return value if value > 0 else fallback


api_key = os.environ.get("UCLOUD_SANDBOX_API_KEY") or os.environ.get("E2B_API_KEY")
domain = normalize_domain(
    os.environ.get("UCLOUD_SANDBOX_DOMAIN")
    or os.environ.get("E2B_DOMAIN")
    or os.environ.get("ASTRAFLOW_SANDBOX_DOMAIN")
    or DEFAULT_DOMAIN
)
cpu_count = read_positive_int("UCLOUD_SANDBOX_TEMPLATE_CPU_COUNT", DEFAULT_CPU_COUNT)
memory_mb = read_positive_int("UCLOUD_SANDBOX_TEMPLATE_MEMORY_MB", DEFAULT_MEMORY_MB)

template = Template().from_template(BASE_TEMPLATE).apt_install(
    ["tmux"], no_install_recommends=True
)

result = Template.build(
    template,
    alias=TEMPLATE_NAME,
    api_key=api_key,
    domain=domain,
    cpu_count=cpu_count,
    memory_mb=memory_mb,
    on_build_logs=default_build_logger(min_level="info"),
)

print(
    json.dumps(
        {
            "name": getattr(result, "name", None),
            "alias": getattr(result, "alias", TEMPLATE_NAME),
            "templateId": getattr(result, "template_id", None)
            or getattr(result, "templateId", None),
            "buildId": getattr(result, "build_id", None)
            or getattr(result, "buildId", None),
            "cpuCount": cpu_count,
            "memoryMB": memory_mb,
        },
        ensure_ascii=False,
        indent=2,
    )
)
